Remove WordPress hooks from the class constructor

jessLundie opened the pull request:
**Changes**

Related to #12 

* Removed the class constructor from each of the plugin classes
* Replaced constructor with an `init()` method that contains all WordPress hooks
* Run method after instantiating the class

This separates out WordPress functionality from the class constructor. In this case, I don't actually believe we need a constructor.

**Testing**

These changes should not alter the function of the plugin in any way. To test:

* Install Plugin Autoupdate Filter on a site with out of date plugins
* When plugin auto-update schedule runs (twice daily run by default), confirm that plugins were updated successfully
* Confirm emails are sent to concierge@ email address

jessLundie committed:
Move hooks out of constructor - main class file

Removes constructor and replaces with an init() for WordPress hooks.

jessLundie committed:
Move hooks out of constructor - self-update file

Remove constructor and move hooks to init() method

jessLundie committed:
Version bump

NickGreen commented on class-plugin-autoupdate-filter.php:
```diff
@@ -12,9 +12,9 @@
+	 * Initialize WordPress hooks
 	 */
-	public function __construct() {
+	public function init() {
```
I believe the is the best way to do it. Could we accomplish something similar with a Singleton Pattern? Yes, probably, but that would also probably be using the pattern for something other than the way it was originally intended. Let's tease this out a little bit.

The most common example for a Singleton Pattern is a database access class, where you only want one instance of that class to exist, because there are potential bad side-effects of having two or more of those objects interacting with the database. 

In our case, what would be the bad side-effects if two instances of this class were instantiated? Not much, really, just some filters being added twice. This might happen in a case of two versions of the same plugin running on the same site. Nothing catastrophic, IMO.

In fact, if we wanted to look more closely at how we are using classes in the plugin, we probably don't even need to use classes. We are using classes to prevent collisions with other code; to separate our code out from any other functions in other custom code or plugins that might use a function of the same name. If we wanted to, we _could_ use namespaces to prevent collision with other code. We'll leave it for now, but something to think about when creating little plugins like this in the future.

NickGreen replied:
Side note: I'm fully aware that I'm the one who originally crafted this plugin using classes, and that I'm bringing my own design choice up for debate 😄

jessLundie replied:
> Can you please figure out the best way to not include the .DS_Store file from your computer in these changes?

Oops! Thank you for catching that. Removed :) 

> If we wanted to, we could use namespaces to prevent collision with other code. We'll leave it for now, but something to think about when creating little plugins like this in the future.

That makes sense. In this case, the class really is just for namespace reasons. That was another question I had while looking at this. Why classes and not namespaces? I know there's been an effort on the Woo side to refactor and switch over to namespaces. 

Related - I couldn't think of any reason someone might want to instantiate a second instance of this class. However, theoretically if there were, that would be an argument for class over namespace. Could there be some benefit from the extensibility side to handling this as a class? Because it makes it more extensible for cases we can't imagine right now?

NickGreen replied:
> Could there be some benefit from the extensibility side to handling this as a class?

Right! That's one great benefit of classes. An example where I've used that extensibility fairly recently is by [extending an Action Scheduler class](https://github.com/a8cteam51/safety-net/blob/trunk/includes/classes/class-actionscheduler-custom-dbstore.php) in order to pause renewal actions. Without the ability to extend that class, our job would've been significantly more difficult.

But, like you said, in this case, with this tiny little plugin which is being distributed mostly only in the context of our own sites we manage, I can't imagine a lot of people needing to extend our classes. In most of these types of cases, we could probably just get away with namespacing it.

NickGreen commented on plugin-autoupdate-filter.php:
```diff
@@ -3,7 +3,7 @@
 Plugin URI: https://github.com/a8cteam51/plugin-autoupdate-filter
 Description: Sets plugin automatic updates to always on, but only happen during specific days and times.
-Version: 1.4.1
+Version: 1.4.2
```
Increasing the version number means we want to craft a new release, which will automatically push out to all of the sites that have this plugin installed (lots).

Does this set of changes warrant a version increase, as described by [SemVer](https://semver.org/)?

jessLundie replied:
🤔 I don't believe that it does. However, SemVer doesn't mention refactoring anywhere.

A quick search led me to [this discussion](https://github.com/semver/semver/issues/146). This argument makes sense to me:

> New bugs may have been introduced, so patch level must be updated

But in our case, if we don't update the patch level, the code won't actually be released until we make further changes to this plugin down the road.

So it seems like to me the question is, "Do we _want_ to release this refactor code and push it to all sites, or do we want to wait and have it included with the next release that merits a version increase per SemVer?"

I do think that we argue for a version bump on the basis plugin isn't updated frequently and doesn't have any kind of regular update schedule. Thoughts?

NickGreen replied:
Thanks for pointing at this discussion, that was a good read. I think in this case, let's not bump the version number or create a new release, and let your refactor get rolled into the next version which contains one or more of these things:
- Bug fix
- Performance increase
- Feature or behavior change for the end user

Also noting here that I did load up this branch on a test site and verified that it works as intended without producing PHP errors in the log.

jessLundie replied:
@NickGreen - I went ahead and reverted the commit that contained the version bump, so we should be all set on this!

NickGreen reviewed (changes requested):
Additional comment: can you please figure out the best way to not include the `.DS_Store` file from your computer in these changes?

Overall, looks good, and opens up for a great discussion!

jessLundie committed:
Removed DS_Store

jessLundie committed:
Revert "Version bump"

This reverts commit 5a6bc69b6b69a9d23745184a4b71fbb8513dd24f.

jessLundie commented:
@NickGreen - I adjusted the version number by reverting the commit, and removed the DS_Store file via [454936f](https://github.com/a8cteam51/plugin-autoupdate-filter/pull/14/commits/454936f38a79e0b5c401b4835c419221d3ae7ff7).

I'm seeing a "Changes requested" flag related to the DS_Store file, but I can't see any way to mark those changes as having been done. Is that something you need to do from your end? 

Once that is cleared up, I think this is okay to merge!

NickGreen commented:
Changes approved. Merge away!
